api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+type ProtocolType string
+
 const (
-	// HTTPProtocol constant.
-	HTTPProtocol = "HTTP"
-	// HTTPSProtocol constant.
-	HTTPSProtocol = "HTTPS"
+	// HTTPProtocolType accepts cleartext HTTP/1.1 sessions over TCP.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// HTTPSProtocolType accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TCPProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
 )
 
-// Listener defines a
+// HostnameMatchType specifies the types of matches that are valid
+// for host names.
+type HostnameMatchType string
+
+const (
+	// HostnameMatchDomain specifies that the host name provided by the
+	// client should be matched against a DNS domain value. The domain
+	// match should follow the wildcard certificate matching rules in
+	// [RFC 6125](https://tools.ietf.org/html/rfc6125#section-6.4.3).
+	HostnameMatchDomain HostnameMatchType = "Domain"
+
+	// HostnameMatchExact specifies that the host name provided by the
+	// client must exactly match the specified value.
+	HostnameMatchExact HostnameMatchType = "Exact"
+)
+
+// HostnameMatch specifies how a Listener should match the host name from a
+// client request. Depending on the incoming protocol, the match may apply to
+// the TLS and the HTTP protocol.
+type HostnameMatch struct {
+	// Match specifies how the host name provided by the client should be
+	// matched against the given value.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=Domain;Exact
+	// +kubebuilder:default=Exact
+	Match HostnameMatchType `json:"match" protobuf:"bytes,1,name=match"`
+
+	// Value contains the value to match against. This value must be a
+	// fully qualified host or domain name as defined by
+	// [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
+	// Note the following deviations from the "host" part of the URI as defined in the RFC:
+	//
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Value string `json:"value" protobuf:"bytes,2,name=value"`
+}
+
+// Listener embodies the concept of a logical endpoint where a
+// Gateway can accept network connections.
+//
+// The GatewayClass may collapse Listener definitions into single
+// implementation-defined acceptor configuration. This allows Gateways to
+// be implemented in terms of virtual services configured within the same
+// proxy server process.
+//
+// Listeners may be considered collapsible if the Protocol field is
+// "HTTP", "HTTPS" or "TLS", the Hostname field is specified, and the Port
+// field matches. As a special case, each group of collapsed Listeners may
+// contain up to one Listener that omits the Hostname field. Listeners that
+// specify identical Hostname fields must not be treated as collapsible.
```

Comment:
> contain up to one Listener that omits the Hostname field.

Why this limitation? 

> Hostname fields must not be treated as collapsible.

> specify identical Hostname fields must not be treated as collapsible.

Which one takes precedence?

Reply:
I updated the text to be a bit more explanatory, but the basic idea is that you have a group of Listeners that you are going to collapse into a single acceptor configuration. Within this group, each Hostname field must be unique (i.e. two Listeners can't match the incoming hostname the same way), and you can only have one default Listener (i.e. only omit the Hostname from one Listener).

Both of these conditions are necessary to avoid ambiguous configurations where a controller would otherwise be required to merge Listeners in some way.

Reply:
https://github.com/kubernetes-sigs/service-apis/pull/193#issuecomment-635665805 explains the gap in my understanding which is why I added the above comment.